src/lib/dialog.js

```diff
@@ -51,49 +53,148 @@ export default class Dialog {
       if (match) {
         return onMatch(match, data);
       }
       return this.choose({
         choices,
         question,
+        prompt,
         exit,
         onMatch,
-        oneTimeHeader: `Sorry, but \`${text}\` is not a valid response. Please try again.`,
+        onError,
+        oneTimeHeader: this._fnOrValue(onError, text, data),
       });
     };
-    return this.ask({
-      exit,
-      message,
+    const options = {
+      question: _question,
       onResponse,
       oneTimeHeader,
+    };
+    if (prompt) { options.prompt = prompt; }
+    if (exit) { options.exit = exit; }
+    return this.ask(options);
+  }
+
+  // Wrapper around single-question methods. Pass in a single question, an
+  // array of questions, or any number of question arguments.
+  //
+  // Questions may be a question object, a function that returns a question
+  // object, or a function that returns a promise that resolves to a question
+  // object.
+  //
+  // Response handlers may return a string to appear as the oneTimeHeader for
+  // the next question, or the final text after all questions. Alternately,
+  // response handlers may return a question object or array of question objects
+  // or a promise that resolves to either.
+  questions(...args) {
+    // Use the first argument if it's an array, otherwise use all arguments.
+    const questions = Array.isArray(args[0]) ? Array.from(args[0]) : args;
+
+    // Determine if the given value is a "question" object. This could perhaps
+    // be more robust.
+    const isQuestion = q => q && q.question;
+
+    // Result might be a promise, so resolve it before anything else.
+    const next = _result => Promise.resolve(_result)
+    .then(result => {
+      // If result is an array of questions or a single question, add them/it
+      // to the beginning of the list.
+      if (Array.isArray(result) && result.every(isQuestion)) {
+        questions.unshift(...result);
+        result = null;
+      }
+      else if (isQuestion(result)) {
+        questions.unshift(result);
+        result = null;
+      }
+      // The next "question" might be a question object, a function that returns
+      // a question object, a promise that resolves to a question object, or a
+      // function that returns a promise that resolves to a question object.
+      return Promise.props({
+        question: questions.length > 0 && this._fnOrValue(questions.shift()),
+        result,
+      });
+    })
+    .then(({question, result}) => {
+      // There's no question. If it's because the question was empty or a
+      // question promise resolved to nothing, get the next question. If it's
+      // because there were no more questions, just return the result.
+      if (!question) {
+        return questions.length === 0 ? result : next(result);
+      }
+      // Guess which method to call, based on the shape of the question object.
+      let apiMethod, responseMethod;
```

Comment:
In these scenarios I personally use this opportunity to define a named thing that describes the conditional. e.g.
```
const questionHasChoices = question.choices
const responseMethod = questionHasChoices ? 'onMatch' : 'onResponse';
```